models/classes/StatisticalMetadata/Import/Builder/ReportBuilder.php

```diff
@@ -0,0 +1,247 @@
+<?php
+
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2022 (original work) Open Assessment Technologies SA.
+ */
+
+declare(strict_types=1);
+
+namespace oat\tao\model\StatisticalMetadata\Import\Builder;
+
+use Throwable;
+use oat\oatbox\reporting\Report;
+use oat\tao\model\StatisticalMetadata\Import\Reporter\ImportReporter;
+use oat\tao\model\StatisticalMetadata\Import\Exception\ErrorValidationException;
+use oat\tao\model\StatisticalMetadata\Import\Exception\HeaderValidationException;
+use oat\tao\model\StatisticalMetadata\Import\Exception\WarningValidationException;
+use oat\tao\model\StatisticalMetadata\Import\Exception\AbstractValidationException;
+
+class ReportBuilder
+{
+    public function buildByReporter(ImportReporter $reporter): Report
+    {
+        $reportType = $this->getReportType($reporter);
+
+        $subReport = $this->createReportByResults($reportType, $reporter);
+        $report = $this
+            ->createReportByResults($reportType, $reporter)
+            ->add($subReport);
+
+        $onlyWarningReports = [];
+        $warningAndErrorReports = [];
+
+        /** @var WarningValidationException[]|ErrorValidationException[] $exceptions */
+        foreach ($reporter->getWarningsAndErrors() as $line => $exceptions) {
+            $headerExceptionReports = [];
+            $warningReports = [];
+            $errorReports = [];
+
+            foreach ($exceptions as $exception) {
+                if ($exception instanceof HeaderValidationException) {
+                    $headerExceptionReports[] = $this->createReportByException($exception);
+
+                    continue;
+                }
+
+                if ($exception instanceof WarningValidationException) {
+                    $warningReports[] = $this->createReportByException($exception);
+
+                    continue;
+                }
+
+                if ($exception instanceof ErrorValidationException) {
+                    $errorReports[] = $this->createReportByException($exception);
+                }
+            }
```

Comment:
Suggestion: This is a simple branching code, I'd use a plain-old `if-else` structure here instead of jumping out of the current loop with `continue`. It makes this nested loop more concise as well. (BTW this is also a personal preference, feel free to keep it as it is :D)

```suggestion
            foreach ($exceptions as $exception) {
                if ($exception instanceof HeaderValidationException) {
                    $headerExceptionReports[] = $this->createReportByException($exception);
                } elseif ($exception instanceof WarningValidationException) {
                    $warningReports[] = $this->createReportByException($exception);
                } elseif ($exception instanceof ErrorValidationException) {
                    $errorReports[] = $this->createReportByException($exception);
                }
            }
```

Reply:
Thanks for your suggestion 🙂
I think, that right now I will keep this one as it is, because this `if-elseif` block make a bit of mess 😅 (I mean, it becomes difficult to read). But I agree with you, that in this case this one is better, than using `continue`.

In the future, I want to refactor and optimize this class 🙂
I will keep your suggestion in my mind to use it during this refactoring